### PR TITLE
Add explanatory note to step 3 in authenticatorMakeCredential

### LINKED ISSUE #1133: add explanatory note to step 3 in authenticatorMakeCredential Operation
I just thought I'd found a bug in http://w3c.github.io/webauthn/#op-make-cred because on initial read it seems that even though  there's a match for an excluded cred, the user is asked for verification and the authnr is supposed to go ahead and create a cred. However, upon searching issues, and reading further in that step to discern that errors are returned in any case, this is all being done for privacy-preserving reasons (see issue #806 )

if we have cycles we ought to put a Note: in that step wrt what's going on...

### PR DESCRIPTION
**Pull request by emlun**
Fixes #1133.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/pull/1326.html" title="Last updated on Oct 21, 2019, 10:31 AM UTC (b58f3a3)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/1326/36fe4ed...b58f3a3.html" title="Last updated on Oct 21, 2019, 10:31 AM UTC (b58f3a3)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -3453,6 +3453,14 @@ When this operation is invoked, the [=authenticator=] MUST perform the following
...
         </dl>
 
+        Note: The purpose of this [=authorization gesture=] is not to proceed with creating a credential,
+        but for privacy reasons to authorize disclosure of the fact that
+        <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code> is [=bound credential|bound=] to this [=authenticator=].
+        If the user consents, the [=client=] and [=[RP]=] can detect this and guide the user to use a different [=authenticator=].
+        If the user does not consent,
+        the [=authenticator=] does not reveal that <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code> is bound to it,
+        and responds as if the user simply declined consent to create a credential.
+
```
Comment (equalsJeffH) [inline index.bs]: the above looks fine, thx!  

Tho, we're saying here: "if the RP gets an `invalidStateError` back from a nav.creds.Create() call, they can do something user-helpful with that info. But this guidance for the RP is buried in "Note:"s in spec sections that we are not suggesting RP devs read.  Additionally, our "rp operations" section only says "abort the ceremony with a user-visible error" if an error is returned from Create().  Seems like we oughta do more.  Maybe we land this and submit a new issue about these other aspects of our authn-hairball-API ?

Comment (emlun) [reply index.bs]: Good comments! A new issue for that seems good to me.

Comment (equalsJeffH) [reply index.bs]: done, see issue #1331

Comment (Robbiesmo) [inline index.bs]: index.bs

Comment (Robbiesmo) [reply index.bs]: 1. If |requireResidentKey| is [TRUE] and the authenticator cannot store a [=client-side-resident public key credential source=]

File: index.bs [hunk 2]
```diff
@@ -3453,6 +3453,14 @@ When this operation is invoked, the [=authenticator=] MUST perform the following
...
 1. If |requireResidentKey| is [TRUE] and the authenticator cannot store a [=client-side-resident public key credential source=],
```
Comment (Robbiesmo) [inline index.bs]: return an error code equivalent to "{{ConstraintError}}" and terminate the operation.
1. If |requireUserVerification| is [TRUE] and the authenticator cannot perform [=user verification=], return an error code

File: index.bs (hunk 1 above)
Comment (emlun) [reply index.bs]: This looks like accidental noise to me. Let me know if it isn't.

File: index.bs (hunk 2 above)
Comment (emlun) [reply index.bs]: This looks like accidental noise to me. Let me know if it isn't.

**Comment by Robbiesmo** [comment]
Op wo 23 okt. 2019 om 12:20 schreef Emil Lundberg <notifications@github.com

**Comment by Robbiesmo** [comment]
diff --git a/index.bs b/index.bs
index a36b66cf..55a2162e 100644
--- a/index.bs
+++ b/index.bs
@@ -3453,6 +3453,15 @@ When this operation is invoked, the [=authenticator=] MUST perform the following
             ::  return an error code equivalent to "{{NotAllowedError}}" and terminate the operation.
         </dl>
 
+        Note: The purpose of this [=authorization gesture=] is not to proceed with creating a credential,
+        but for privacy reasons to authorize disclosure of the fact that
+        <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code> is [=bound credential|bound=] to this [=authenticator=].
+        If the user consents, the [=client=] and [=[RP]=] can detect this and guide the user to use a different [=authenticator=].
+        If the user does not consent,
+        the [=authenticator=] does not reveal that <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code>
+        is [=bound credential|bound=] to it,
+        and responds as if the user simply declined consent to create a credential.
+
 1. If |requireResidentKey| is [TRUE] and the authenticator cannot store a [=client-side-resident public key credential source=],
     return an error code equivalent to "{{ConstraintError}}" and terminate the operation.
 1. If |requireUserVerification| is [TRUE] and the authenticator cannot perform [=user verification=], return an error code

File: index.bs (hunk 2 above)
Comment (Robbiesmo) [inline index.bs]: diff --git a/index.bs b/index.bs
index a36b66cf..55a2162e 100644
--- a/index.bs
+++ b/index.bs
@@ -3453,6 +3453,15 @@ When this operation is invoked, the [=authenticator=] MUST perform the following
             ::  return an error code equivalent to "{{NotAllowedError}}" and terminate the operation.
         </dl>
 
+        Note: The purpose of this [=authorization gesture=] is not to proceed with creating a credential,
+        but for privacy reasons to authorize disclosure of the fact that
+        <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code> is [=bound credential|bound=] to this [=authenticator=].
+        If the user consents, the [=client=] and [=[RP]=] can detect this and guide the user to use a different [=authenticator=].
+        If the user does not consent,
+        the [=authenticator=] does not reveal that <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code>
+        is [=bound credential|bound=] to it,
+        and responds as if the user simply declined consent to create a credential.
+
 1. If |requireResidentKey| is [TRUE] and the authenticator cannot store a [=client-side-resident public key credential source=],
     return an error code equivalent to "{{ConstraintError}}" and terminate the operation.
 1. If |requireUserVerification| is [TRUE] and the authenticator cannot perform [=user verification=], return an error code

Comment (Robbiesmo) [reply index.bs]: From 50a22e4bb3730c1e83f2503c52511eaedffb2a0e Mon Sep 17 00:00:00 2001
From: Emil Lundberg <emil@yubico.com>
Date: Fri, 18 Oct 2019 16:58:23 +0200
Subject: [PATCH 1/2] Add explanatory note to step 3 in
 authenticatorMakeCredential

---
 index.bs | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/index.bs b/index.bs
index a36b66cf..a6fea5a4 100644
--- a/index.bs
+++ b/index.bs
@@ -3453,6 +3453,14 @@ When this operation is invoked, the [=authenticator=] MUST perform the following
             ::  return an error code equivalent to "{{NotAllowedError}}" and terminate the operation.
         </dl>
 
+        Note: The purpose of this [=authorization gesture=] is not to proceed with creating a credential,
+        but for privacy reasons to authorize disclosure of the fact that
+        <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code> is [=bound credential|bound=] to this [=authenticator=].
+        If the user consents, the [=client=] and [=[RP]=] can detect this and guide the user to use a different [=authenticator=].
+        If the user does not consent,
+        the [=authenticator=] does not reveal that <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code> is bound to it,
+        and responds as if the user simply declined consent to create a credential.
+
 1. If |requireResidentKey| is [TRUE] and the authenticator cannot store a [=client-side-resident public key credential source=],
     return an error code equivalent to "{{ConstraintError}}" and terminate the operation.
 1. If |requireUserVerification| is [TRUE] and the authenticator cannot perform [=user verification=], return an error code

From b58f3a308be5d26e7401ca7813da2ccfebd17d47 Mon Sep 17 00:00:00 2001
From: Emil Lundberg <emil@yubico.com>
Date: Mon, 21 Oct 2019 12:31:00 +0200
Subject: [PATCH 2/2] Add definition link to "bound" term

---
 index.bs | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/index.bs b/index.bs
index a6fea5a4..55a2162e 100644
--- a/index.bs
+++ b/index.bs
@@ -3458,7 +3458,8 @@ When this operation is invoked, the [=authenticator=] MUST perform the following
         <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code> is [=bound credential|bound=] to this [=authenticator=].
         If the user consents, the [=client=] and [=[RP]=] can detect this and guide the user to use a different [=authenticator=].
         If the user does not consent,
-        the [=authenticator=] does not reveal that <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code> is bound to it,
+        the [=authenticator=] does not reveal that <code>|descriptor|.{{PublicKeyCredentialDescriptor/id}}</code>
+        is [=bound credential|bound=] to it,
         and responds as if the user simply declined consent to create a credential.
 
 1. If |requireResidentKey| is [TRUE] and the authenticator cannot store a [=client-side-resident public key credential source=],

Comment (emlun) [reply index.bs]: This looks like accidental noise to me. Let me know if it isn't.

**Comment by holycleugh** [comment]
Hi, I'm trying to resolve a possible conflict between this and the CTAP2 (Jan 30, 2019). Let me know if there's a more appropriate forum to post this question on.

In the WebAuthn algorithm, if the credential is excluded and the user declines consent, we return NotAllowedError. However, the CTAP2 spec says

>5.1
>If the excludeList parameter is present and contains a credential ID that is present on this authenticator and bound to the specified rpId, wait for user presence, then terminate this procedure and return error code CTAP2_ERR_CREDENTIAL_EXCLUDED.

It's possibly ambiguous but it seems to return CTAP2_ERR_CREDENTIAL_EXCLUDED regardless of whether the user gave consent. Should I treat the new WebAuthn guidelines as superseding the older CTAP2 spec and return CTAP2_ERR_OPERATION_DENIED if the user declined consent?

**Comment by emlun** [comment]
I think CTAP2 is intended to say to return CTAP2_ERR_OPERATION_DENIED if the user presence check fails, and I believe that's what authenticators do, but I agree this is ambiguous in CTAP2.

**Pull request merged**